File: built-in plugins dsp/mb-chorus-dimension.cpp
/**
 * MB Dimension
 * Category : effect
 * Type     : chorus
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Dimension D style
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_CHORUS_DIMENSION_H
#define MB_CHORUS_DIMENSION_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbChorusDimension : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-chorus-dimension";
    static constexpr const char* PLUGIN_NAME    = "MB Dimension";
    static constexpr const char* PLUGIN_TYPE    = "chorus";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float mode = 2f;  // range [1, 4]
    float mix = 0.5f;  // range [0, 1]
    };

    MbChorusDimension() = default;
    ~MbChorusDimension() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.mode = std::clamp(params.mode, 1f, 4f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Dimension
        return input;
    }
};

#endif // MB_CHORUS_DIMENSION_H

File: built-in plugins dsp/mb-chorus-dimension.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbChorusDimensionPlugin: PluginDefinition = { id: 'mb-chorus-dimension', slug: 'mb-chorus-dimension', name: 'MB Dimension', category: 'effect', type: 'chorus', version: '1.0.0', description: 'Dimension D style', author: 'Max Booster', grade: 'A', parameters: [{ id: 'mode', name: 'Mode', type: 'float', defaultValue: 2, minValue: 1, maxValue: 4, automatable: false }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { mode: 2, mix: 0.5 } };

export default MbChorusDimensionPlugin;
